src-java/kilda-model/src/main/java/org/openkilda/model/HaFlow.java

```diff
@@ -0,0 +1,593 @@
+        if (getReversePathId() == null) {
+            return null;
+        }
+        return getPath(getReversePathId()).orElse(null);
+    }
+
+
+    /**
+     * Add a path and set it as the reverse path.
+     */
+    public void setReversePath(HaFlowPath reversePath) {
+        if (reversePath != null) {
+            if (!hasPath(reversePath)) {
+                validateReversePath(reversePath);
+                addPaths(reversePath);
+            }
+            setReversePathId(reversePath.getHaPathId());
+        } else {
+            setReversePathId(null);
+        }
+    }
+
+    /**
+     * Get the protected reverse path.
+     */
+    public HaFlowPath getProtectedReversePath() {
+        if (getProtectedReversePathId() == null) {
+            return null;
+        }
+        return getPath(getProtectedReversePathId()).orElse(null);
+    }
+
+    /**
+     * Add a path and set it as the protected reverse.
+     */
+    public void setProtectedReversePath(HaFlowPath reversePath) {
+        if (reversePath != null) {
+            if (!hasPath(reversePath)) {
+                validateReversePath(reversePath);
+                addPaths(reversePath);
+            }
+            setProtectedReversePathId(reversePath.getHaPathId());
+        } else {
+            setProtectedReversePathId(null);
+        }
+    }
+
+    private void validateEndpoints(HaFlowPath path) {
+        if (!path.getSharedSwitchId().equals(getSharedEndpoint().switchId)) {
+            throw new IllegalArgumentException(format("HA-path %s has shared endpoint switch id %s, but %s expected",
+                    path.getHaPathId(), path.getSharedSwitchId(), getSharedEndpoint().switchId));
+        }
+        Set<SwitchId> subFlowSwitchIds = getSubFlows().stream()
+                .map(HaSubFlow::getEndpointSwitchId)
+                .collect(Collectors.toSet());
+        Set<SwitchId> pathSwitchIds = path.getSubFlowSwitchIds();
+
+        if (!subFlowSwitchIds.equals(pathSwitchIds)) {
+            throw new IllegalArgumentException(format("HA-path %s must have %s switch endpoints, but has following "
+                            + "endpoints: %s",
+                    path.getHaPathId(), subFlowSwitchIds, pathSwitchIds));
+        }
+    }
+
+    private void validateForwardPath(HaFlowPath haPath) {
+        validateEndpoints(haPath);
+        if (!haPath.isForward()) {
+            throw new IllegalArgumentException(format("Cookie %s of HA-path %s must have forward direction",
+                    haPath.getCookie(), haPath.getHaPathId()));
+        }
+    }
+
+    private void validateReversePath(HaFlowPath haPath) {
+        validateEndpoints(haPath);
+        if (!haPath.isReverse()) {
+            throw new IllegalArgumentException(format("Cookie %s of HA-path %s must have reverse direction",
+                    haPath.getCookie(), haPath.getHaPathId()));
+        }
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
out of curiosity: why not to use the annotations to generate equals and hashCode?

Reply:
Db objects are tricky. Each object has inner field `data` which contains actual data fields. It we will use @EqualsAndHashCode it will compare whole `data` object. But we need to compare a set of data fields. Also `data` is an interface which has 2 implementations: POJO and object, connected with DB. Both implementation must have equal `equals()` so to do not duplicate equal and hashcode for each implementation we can add `equal()` here manually.